README
vatnumber
=========

Python module to validate VAT numbers.

Nutshell
--------

Here a simple example to validate VAT numbers format::

    >>> import vatnumber
    >>> vatnumber.check_vat('BE0123456749')
    True

Here a simple example to validate European VAT through VIES service::

    >>> import vatnumber
    >>> vatnumber.check_vies('BE0897290877')
    True


For more information please visit the `vatnumber website_`.

.. _vatnumber website: http://code.google.com/p/vatnumber/


CHANGELOG
Version 1.2 - 2014-06-25
* Use python-stdnum
* Add support for proxy

Version 1.1 - 2013-06-01
* Add Python 3 support
* Add Check Argentina VAT number
* Fix check for negative integer

Version 1.0 - 2011-09-28
* Fix UK VAT check
* Fix UK VAT 9755 validation
* Fix Sweden 2 last numbers validation
* Add 999 and 888 as valide IT province of residence

Version 0.9 - 2011-03-17
* Fix VIES URL

Version 0.8 - 2011-01-18
* Add GB validation for the "9755" checksum introduced in November 2009

Version 0.7 - 2010-08-24
* Fix test on last char of Albania
* Add Check Croatia VAT number

Version 0.6 - 2010-05-30
* Add generic check_vat method
* Replace SOAPpy by suds
* Add GB validation for {GB,HA}888 8xxx yy format
* Add GB validation for branch traders 12 digits
* Update vies SOAP URL and add unittest
* Add Check Russia VAT number

Version 0.5 - 2009-11-13
* Add Check Chile RUT number for issue5
* Add Check Colombian RUT number for issue4
* Fix Portugal VAT when check number is 11

Version 0.4 - 2009-03-30
* Update Spain test for issue3 thanks to carlos@pemas.es
* Add VIES check with the SOAP Service of EU
* Add test for Bulgaria VAT number with length of 9

Version 0.3 - 2009-03-23
* Fix check_vat_ro for vat number with less than 10 digits

Version 0.2 - 2008-09-29
* Add check_vat_uk function
* Improve Belgium vat check as first number must be 0
* Add Check Albania VAT number
* Add Check San Marino VAT number
* Add Check Ukraine VAT number

Version 0.1 - 2008-09-20
* Initial release


tests.py
#This file is part of vatnumber.  The COPYRIGHT file at the top level of
#this repository contains the full copyright notices and license terms.
'''
Unit test for vatnumber
'''

import unittest
import vatnumber

VAT_NUMBERS = [
    ('AT', 'U12345675'),
    ('AL', 'K99999999L'),
    ('BE', '0123456749'),
    ('BG', '1234567892'),
    ('BG', '175074752'),
    ('BG', '131202360'),
    ('BG', '040683212'),
    ('CO', '9001279338'),
    ('CY', '12345678F'),
    ('CZ', '12345679'),
    ('CZ', '612345670'),
    ('CZ', '991231123'),
    ('CZ', '6306150004'),
    ('DE', '123456788'),
    ('DK', '12345674'),
    ('EE', '123456780'),
    ('ES', 'A12345674'),
    ('ES', 'P1234567D'),
    ('ES', 'K1234567L'),
    ('ES', 'R9600075G'),
    ('ES', 'W4003922D'),
    ('ES', 'V99218067'),
    ('ES', 'U99216632'),
    ('ES', 'J99216582'),
    ('ES', 'U99216426'),
    ('ES', '12345678Z'),
    ('ES', 'X5277343Q'),
    ('ES', 'Y5277343F'),
    ('ES', 'Z5277343K'),
    ('FI', '12345671'),
    ('FR', '32123456789'),
    ('FR', '2H123456789'),
    ('GB', 'GD123'),
    ('GB', 'HA567'),
    ('GB', '123456782'),
    ('GB', '1234567823'),
    ('GB', '001123456782'),
    ('GB', '0011234567823'),
    ('GR', '12345670'),
    ('GR', '123456783'),
    ('HU', '12345676'),
    ('IE', '7A12345J'),
    ('IE', '1234567T'),
    ('IT', '12345670017'),
    ('LT', '123456715'),
    ('LT', '123456789011'),
    ('LU', '12345613'),
    ('LV', '41234567891'),
    ('LV', '15066312345'),
    ('MT', '12345634'),
    ('NL', '123456782B90'),
    ('PL', '1234567883'),
    ('PT', '123456789'),
    ('RO', '24736200'),
    ('RO', '1234567897'),
    ('RO', '1630615123457'),
    ('SE', '123456789701'),
    ('SI', '12345679'),
    ('SK', '0012345675'),
    ('SK', '0012345678'),
    ('SK', '531231123'),
    ('SK', '6306151234'),
    ('SM', '12345'),
    ('UA', '12345678'),
]


class VatNumberTest(unittest.TestCase):
    '''
    Test Case for vatnumber
    '''

    def test_vat_numbers(self):
        '''
        Test VAT numbers
        '''
        for code, number in VAT_NUMBERS:
            self.assert_(getattr(vatnumber,
                'check_vat_' + code.lower())(number), code + ' ' + number)

if __name__ == '__main__':
    unittest.main()


vatnumber/tests.py
#This file is part of vatnumber.  The COPYRIGHT file at the top level of
#this repository contains the full copyright notices and license terms.
'''
Unit test for vatnumber
'''

import unittest
import vatnumber

VAT_NUMBERS = [
    ('AT', 'U12345675', True),
    ('AT', 'U123456789', False),
    ('AT', 'A12345675', False),
    ('AT', 'UA2345675', False),
    ('AT', 'U12345620', True),
    ('AT', 'U12345678', False),
    ('AL', 'K99999999L', True),
    ('AL', 'K999999999L', False),
    ('AL', 'AA9999999L', False),
    ('AL', 'KA9999999L', False),
    ('AL', 'K999999991', False),
    ('AR', '00000000000', True),
    ('AR', '00000000001', False),
    ('BE', '0123456749', True),
    ('BE', '0897290877', True),
    ('BE', '01234567490', False),
    ('BE', '9123456749', False),
    ('BE', '0A23456749', False),
    ('BE', '0123456700', False),
    ('BG', '1234567892', True),
    ('BG', '175074752', True),
    ('BG', '131202360', True),
    ('BG', '040683212', True),
    ('BG', '12345678921', False),
    ('BG', 'A234567892', False),
    ('BG', '2234567892', False),
    ('BG', '1001000000', True),
    ('BG', '0000003000', False),
    ('BG', '1234567890', False),
    ('CL', '334441113', True),
    ('CL', 'A34441113', False),
    ('CL', '334441180', True),
    ('CL', '33444113K', True),
    ('CO', '9001279338', True),
    ('CO', '900127933', False),
    ('CO', 'A001279338', False),
    ('CO', '9001279320', True),
    ('CY', '12345678F', True),
    ('CY', '2345678F', False),
    ('CY', 'A2345678F', False),
    ('CY', '12345678A', False),
    ('CZ', '1234567', False),
    ('CZ', '12345679', True),
    ('CZ', 'A2345679', False),
    ('CZ', '92345679', False),
    ('CZ', '10001000', True),
    ('CZ', '10000101', True),
    ('CZ', '12345670', False),
    ('CZ', '612345670', True),
    ('CZ', '612345679', False),
    ('CZ', '991231123', True),
    ('CZ', '541231123', False),
    ('CZ', '791231123', False),
    ('CZ', '990031123', False),
    ('CZ', '991331123', False),
    ('CZ', '995031123', False),
    ('CZ', '996331123', False),
    ('CZ', '990200123', False),
    ('CZ', '995229123', False),
    ('CZ', '965200123', False),
    ('CZ', '960230123', False),
    ('CZ', '990400123', False),
    ('CZ', '990431123', False),
    ('CZ', '990100123', False),
    ('CZ', '990132123', False),
    ('CZ', '6306150004', True),
    ('CZ', '5306150004', False),
    ('CZ', '6300150004', False),
    ('CZ', '6313150004', False),
    ('CZ', '6350150004', False),
    ('CZ', '6363150004', False),
    ('CZ', '6302000004', False),
    ('CZ', '6302290004', False),
    ('CZ', '6402000004', False),
    ('CZ', '6402310004', False),
    ('CZ', '6304000004', False),
    ('CZ', '6304310004', False),
    ('CZ', '6301000004', False),
    ('CZ', '6301320004', False),
    ('CZ', '6306150000', False),
    ('CZ', '6306150004', True),
    ('DE', '123456788', True),
    ('DE', '12345678', False),
    ('DE', 'A23456788', False),
    ('DE', '000000088', False),
    ('DE', '123456770', True),
    ('DE', '123456789', False),
    ('DK', '12345674', True),
    ('DK', '1234564', False),
    ('DK', 'A2345674', False),
    ('DK', '02345674', False),
    ('DK', '12345679', False),
    ('EE', '123456780', True),
    ('EE', '1234567890', False),
    ('EE', 'A23456780', False),
    ('EE', '123456789', False),
    ('ES', 'A12345674', True),
    ('ES', 'P1234567D', True),
    ('ES', 'K1234567L', True),
    ('ES', 'R9600075G', True),
    ('ES', 'W4003922D', True),
    ('ES', 'V99218067', True),
    ('ES', 'U99216632', True),
    ('ES', 'J99216582', True),
    ('ES', 'U99216426', True),
    ('ES', '12345678Z', True),
    ('ES', 'X5277343Q', True),
    ('ES', 'Y5277343F', True),
    ('ES', 'Z5277343K', True),
    ('ES', '1234567890', False),
    ('ES', 'AB3456789', False),
    ('ES', 'A12345690', True),
    ('ES', 'A12345679', False),
    ('ES', 'WA003922D', False),
    ('ES', 'W4003922A', False),
    ('ES', 'ZA277343K', False),
    ('ES', 'Z5277343A', False),
    ('ES', '1A345678Z', False),
    ('ES', '12345678A', False),
    ('FI', '12345671', True),
    ('FR', '32123456789', True),
    ('FR', '2H123456789', True),
    ('GB', 'GD123', True),
    ('GB', 'GD888812326', True),
    ('GB', 'HA567', True),
    ('GB', 'HA888856782', True),
    ('GB', '123456782', True),
    ('GB', '102675046', True),
    ('GB', '100190874', True),
    ('GB', '003232345', True),
    ('GB', '1234567823', True),
    ('GB', '001123456782', True),
    ('GB', '0011234567823', True),
    ('GB', '242338087388', True),
    ('GR', '12345670', True),
    ('GR', '123456783', True),
    ('HR', '12345678903', True),
    ('HR', '24595836665', True),
    ('HR', '23448731483', True),
    ('HU', '12345676', True),
    ('IE', '7A12345J', True),
    ('IE', '1234567T', True),
    ('IT', '12345670017', True),
    ('IT', '00118439991', True),
    ('LT', '123456715', True),
    ('LT', '123456789011', True),
    ('LU', '12345613', True),
    ('LV', '41234567891', True),
    ('LV', '15066312345', True),
    ('MT', '12345634', True),
    ('NL', '123456782B90', True),
    ('PL', '1234567883', True),
    ('PT', '123456789', True),
    ('RO', '24736200', True),
    ('RO', '1234567897', True),
    ('RO', '1630615123457', True),
    ('RO', '-7793957', False),
    ('RU', '5505035011', True),
    ('RU', '550501929014', True),
    ('SE', '123456789701', True),
    ('SE', '556728341001', True),
    ('SI', '12345679', True),
    ('SK', '0012345675', True),
    ('SK', '0012345678', True),
    ('SK', '531231123', True),
    ('SK', '6306151234', True),
    ('SK', '2021853504', True),
    ('SM', '12345', True),
    ('UA', '12345678', True),
    ('', '12456789', False),
]

VIES_NUMBERS = [
    'BE0897290877',
]


class VatNumberTest(unittest.TestCase):
    '''
    Test Case for vatnumber
    '''

    def test_vat_numbers(self):
        '''
        Test VAT numbers
        '''
        for code, number, result in VAT_NUMBERS:
            if result:
                test = self.assertTrue
            else:
                test = self.assertFalse
            test(vatnumber.check_vat(code + number), code + number)

    def test_vies(self):
        '''
        Test vies
        '''
        for vat in VIES_NUMBERS:
            self.assert_(vatnumber.check_vies(vat))

    def test_countries(self):
        '''
        Test countries
        '''
        vatnumber.countries()

if __name__ == '__main__':
    unittest.main()


vatnumber/__init__.py
#This file is part of vatnumber.  The COPYRIGHT file at the top level of
#this repository contains the full copyright notices and license terms.
'''
Check the VAT number depending of the country based on formula from
http://sima-pc.com/nif.php (dead link, was http://web.archive.org/web/20120118023804/http://sima-pc.com/nif.php)
http://en.wikipedia.org/wiki/Vat_number
'''
__version__ = '1.3'


def _posint(x):
    value = int(x)
    if value < 0:
        raise ValueError('not a positive integer')
    return value


def countries():
    '''
    Return the list of country's codes that have check function
    '''
    res = [x.replace('check_vat_', '').upper() for x in globals()
        if x.startswith('check_vat_')]
    res.sort()
    return res


def mult_add(i, j):
    '''
    Sum each digits of the multiplication of i and j.
    '''
    mult = i * j
    res = 0
    for i in range(len(str(mult))):
        res += int(str(mult)[i])
    return res


def mod1110(value):
    '''
    Compute ISO 7064, Mod 11,10
    '''
    t = 10
    for i in value:
        c = int(i)
        t = (2 * ((t + c) % 10 or 10)) % 11
    return (11 - t) % 10


def check_vat_at(vat):
    '''
    Check Austria VAT number.
    '''
    import stdnum.at.uid
    return stdnum.at.uid.is_valid(vat)


def check_vat_al(vat):
    '''
    Check Albania VAT number.
    '''
    if len(vat) != 10:
        return False
    if vat[0] not in ('J', 'K'):
        return False
    try:
        _posint(vat[1:9])
    except ValueError:
        return False
    if ord(vat[9]) < 65 or ord(vat[9]) > 90:
        return False
    return True


def check_vat_ar(vat):
    '''
    Check Argentina VAT number.
    '''
    if len(vat) != 11:
        return False
    base = [5, 4, 3, 2, 7, 6, 5, 4, 3, 2]
    aux = 0
    for i in xrange(10):
        aux += int(vat[i]) * base[i]
    aux = 11 - (aux - (int(aux / 11) * 11))
    if aux == 11:
        aux = 0
    if aux == 10:
        aux = 9
    return aux == int(vat[10])


def check_vat_be(vat):
    '''
    Check Belgium VAT number.
    '''
    import stdnum.be.vat
    return stdnum.be.vat.is_valid(vat)


def check_vat_bg(vat):
    '''
    Check Bulgaria VAT number.
    '''
    import stdnum.bg.vat
    return stdnum.bg.vat.is_valid(vat)


def check_vat_cl(rut):
    '''
    Check Chile RUT number.
    '''
    try:
        _posint(rut[:-1])
    except ValueError:
        return False
    sum = 0
    for i in range(len(rut) - 2, -1, -1):
        sum += int(rut[i]) * (((len(rut) - 2 - i) % 6) + 2)
    check = 11 - (sum % 11)
    if check == 11:
        return rut[-1] == '0'
    elif check == 10:
        return rut[-1].upper() == 'K'
    else:
        return check == int(rut[-1])


def check_vat_co(rut):
    '''
    Check Colombian RUT number.
    '''
    if len(rut) != 10:
        return False
    try:
        _posint(rut)
    except ValueError:
        return False
    nums = [3, 7, 13, 17, 19, 23, 29, 37, 41, 43, 47, 53, 59, 67, 71]
    sum = 0
    for i in range(len(rut) - 2, -1, -1):
        sum += int(rut[i]) * nums[len(rut) - 2 - i]
    if sum % 11 > 1:
        return int(rut[-1]) == 11 - (sum % 11)
    else:
        return int(rut[-1]) == sum % 11


def check_vat_cy(vat):
    '''
    Check Cyprus VAT number.
    '''
    import stdnum.cy.vat
    return stdnum.cy.vat.is_valid(vat)


def check_vat_cz(vat):
    '''
    Check Czech Republic VAT number.
    '''
    import stdnum.cz.dic
    return stdnum.cz.dic.is_valid(vat)


def check_vat_de(vat):
    '''
    Check Germany VAT number.
    '''
    import stdnum.de.vat
    return stdnum.de.vat.is_valid(vat)


def check_vat_dk(vat):
    '''
    Check Denmark VAT number.
    '''
    import stdnum.dk.cvr
    return stdnum.dk.cvr.is_valid(vat)


def check_vat_ee(vat):
    '''
    Check Estonia VAT number.
    '''
    import stdnum.ee.kmkr
    return stdnum.ee.kmkr.is_valid(vat)


def check_vat_es(vat):
    '''
    Check Spain VAT number.
    '''
    import stdnum.es.nif
    return stdnum.es.nif.is_valid(vat)


def check_vat_fi(vat):
    '''
    Check Finland VAT number.
    '''
    import stdnum.fi.alv
    return stdnum.fi.alv.is_valid(vat)


def check_vat_fr(vat):
    '''
    Check France VAT number.
    '''
    import stdnum.fr.tva
    return stdnum.fr.tva.is_valid(vat)


def check_vat_gb(vat):
    '''
    Check United Kingdom VAT number.
    '''
    import stdnum.gb.vat
    return stdnum.gb.vat.is_valid(vat)


def check_vat_gr(vat):
    '''
    Check Greece VAT number.
    '''
    import stdnum.gr.vat
    return stdnum.gr.vat.is_valid(vat)


def check_vat_el(vat):
    '''
    Check Greece VAT number.
    '''
    return check_vat_gr(vat)


def check_vat_hr(vat):
    '''
    Check Croatia VAT number.
    '''
    import stdnum.hr.oib
    return stdnum.hr.oib.is_valid(vat)


def check_vat_hu(vat):
    '''
    Check Hungary VAT number.
    '''
    import stdnum.hu.anum
    return stdnum.hu.anum.is_valid(vat)


def check_vat_ie(vat):
    '''
    Check Ireland VAT number.
    '''
    import stdnum.ie.vat
    return stdnum.ie.vat.is_valid(vat)


def check_vat_it(vat):
    '''
    Check Italy VAT number.
    '''
    import stdnum.it.iva
    return stdnum.it.iva.is_valid(vat)


def check_vat_lt(vat):
    '''
    Check Lithuania VAT number.
    '''
    import stdnum.lt.pvm
    return stdnum.lt.pvm.is_valid(vat)


def check_vat_lu(vat):
    '''
    Check Luxembourg VAT number.
    '''
    import stdnum.lu.tva
    return stdnum.lu.tva.is_valid(vat)


def check_vat_lv(vat):
    '''
    Check Latvia VAT number.
    '''
    import stdnum.lv.pvn
    return stdnum.lv.pvn.is_valid(vat)


def check_vat_mt(vat):
    '''
    Check Malta VAT number.
    '''
    import stdnum.mt.vat
    return stdnum.mt.vat.is_valid(vat)


def check_vat_nl(vat):
    '''
    Check Netherlands VAT number.
    '''
    import stdnum.nl.btw
    return stdnum.nl.btw.is_valid(vat)


def check_vat_pl(vat):
    '''
    Check Poland VAT number.
    '''
    import stdnum.pl.nip
    return stdnum.pl.nip.is_valid(vat)


def check_vat_pt(vat):
    '''
    Check Portugal VAT number.
    '''
    import stdnum.pt.nif
    return stdnum.pt.nif.is_valid(vat)


def check_vat_ro(vat):
    '''
    Check Romania VAT number.
    '''
    import stdnum.ro.cf
    return stdnum.ro.cf.is_valid(vat)


def check_vat_se(vat):
    '''
    Check Sweden VAT number.
    '''
    import stdnum.se.vat
    return stdnum.se.vat.is_valid(vat)


def check_vat_si(vat):
    '''
    Check Slovenia VAT number.
    '''
    import stdnum.si.ddv
    return stdnum.si.ddv.is_valid(vat)


def check_vat_sk(vat):
    '''
    Check Slovakia VAT number.
    '''
    import stdnum.sk.dph
    return stdnum.sk.dph.is_valid(vat)


def check_vat_sm(vat):
    '''
    Check San Marino VAT number.
    '''
    if len(vat) != 5:
        return False
    try:
        _posint(vat)
    except ValueError:
        return False
    return True


def check_vat_ua(vat):
    '''
    Check Ukraine VAT number.
    '''
    if len(vat) != 8:
        return False
    try:
        _posint(vat)
    except ValueError:
        return False
    return True


def check_vat_uk(vat):
    '''
    Check United Kingdom VAT number.
    '''
    return check_vat_gb(vat)


def check_vat_ru(vat):
    '''
    Check Russia VAT number.
    '''
    if len(vat) != 10 and len(vat) != 12:
        return False
    try:
        _posint(vat)
    except ValueError:
        return False

    if len(vat) == 10:
        check_sum = 2 * int(vat[0]) + 4 * int(vat[1]) + 10 * int(vat[2]) + \
            3 * int(vat[3]) + 5 * int(vat[4]) + 9 * int(vat[5]) + \
            4 * int(vat[6]) + 6 * int(vat[7]) + 8 * int(vat[8])
        check = check_sum % 11
        if check % 10 != int(vat[9]):
            return False
    else:
        check_sum1 = 7 * int(vat[0]) + 2 * int(vat[1]) + 4 * int(vat[2]) + \
            10 * int(vat[3]) + 3 * int(vat[4]) + 5 * int(vat[5]) + \
            9 * int(vat[6]) + 4 * int(vat[7]) + 6 * int(vat[8]) + \
            8 * int(vat[9])
        check = check_sum1 % 11

        if check != int(vat[10]):
            return False
        check_sum2 = 3 * int(vat[0]) + 7 * int(vat[1]) + 2 * int(vat[2]) + \
            4 * int(vat[3]) + 10 * int(vat[4]) + 3 * int(vat[5]) + \
            5 * int(vat[6]) + 9 * int(vat[7]) + 4 * int(vat[8]) + \
            6 * int(vat[9]) + 8 * int(vat[10])
        check = check_sum2 % 11
        if check != int(vat[11]):
            return False
    return True


def check_vat(vat):
    '''
    Check VAT number.
    '''
    code = vat[:2].lower()
    number = vat[2:]
    try:
        checker = globals()['check_vat_%s' % code]
    except KeyError:
        return False
    return checker(number)


def check_vies(vat):
    '''
    Check VAT number for EU member state using the SOAP Service
    '''
    import stdnum.eu.vat
    return bool(stdnum.eu.vat.check_vies(vat)['valid'])
